File: 62-unique-paths/unique-paths.cpp
class Solution {
public:
    int helper(int x, int y, vector<vector<int>>& dp){
        if(x == 0 && y == 0) return 1;
        if(x < 0 || y < 0){
            return 0;
        }

        if(dp[x][y] != -1) return dp[x][y];

        int left = 0;
        int up = 0;

        left = helper(x, y-1, dp);
        up = helper(x-1, y, dp);

        return dp[x][y] = left + up;
    }
    int uniquePaths(int m, int n) {
        vector<vector<int>>dp(m, vector<int>(n, -1));
        return helper(m-1, n-1, dp);
    }
};

File: 62-unique-paths/unique-paths.py
class Solution:
    def helper(self, i: int, j: int, dp: List[List[int]]) -> int:
        if i < 0 or j < 0: 
            return 0
        if i == 0 and j == 0:
            return 1

        if dp[i][j] != -1:
            return dp[i][j]
        
        up = self.helper(i-1, j, dp)
        left = self.helper(i, j-1, dp)

        dp[i][j] = up + left

        return dp[i][j]
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[-1 for _ in range(n)] for _ in range(m)]
        return self.helper(m-1, n-1, dp)